File: XUCore/additions/NSWindow-NoodleEffects.m
// 
// NSWindow-NoodleEffects.m
// 
// Created by Charlie Monroe
// 
// Copyright (c) 2010-13 Fuel Collective, LLC. All rights reserved.
// 


#import "NSWindow-NoodleEffects.h"
#import <Carbon/Carbon.h>

#define ZOOM_ANIMATION_TIME_MULTIPLIER			0.4
#define SLOW_ZOOM_ANIMATION_TIME_MULTIPLIER		5

@interface NoodleZoomWindow : NSPanel

@end

@implementation NoodleZoomWindow

- (NSTimeInterval)animationResizeTime:(NSRect)newWindowFrame
{
	float			multiplier;
	
	multiplier = ZOOM_ANIMATION_TIME_MULTIPLIER;


	/* if (GetCurrentEventKeyModifiers() & shiftKey)
	{

		multiplier = SLOW_ZOOM_ANIMATION_TIME_MULTIPLIER;
	}
	 */
	
	return [super animationResizeTime:newWindowFrame] * multiplier;
}

@end

@implementation NSWindow (NoodleEffects)

static NSPanel *__zoomWindow;

- (NSPanel *)_createZoomWindowWithRect:(NSRect)rect
{
	NSPanel *zoomWindow;
	NSImageView     *imageView;
	NSImage         *image;
	NSRect          frame;
	BOOL            isOneShot;
	
	frame = [self frame];

	isOneShot = [self isOneShot];
	if (isOneShot)
	{
		[self setOneShot:NO];
	}
	
	if ([self windowNumber] <= 0)
	{
		// Force window device. Kinda crufty but I don't see a visible flash
		// when doing this. May be a timing thing wrt the vertical refresh.
		[self orderBack:self];
		[self orderOut:self];
	}
	
	/* [self orderBack:self];
		[self orderOut:self]; */
	
		
	image = [[NSImage alloc] initWithSize:frame.size];
	
	NSView *view = [[self contentView] superview];
	NSBitmapImageRep *imageRep = [view bitmapImageRepForCachingDisplayInRect:[view bounds]];
	[view cacheDisplayInRect:[view bounds] toBitmapImageRep:imageRep];
	[image addRepresentation:imageRep];
	
	zoomWindow = [[NoodleZoomWindow alloc] initWithContentRect:rect styleMask:NSBorderlessWindowMask backing:NSBackingStoreBuffered defer:YES];
	[zoomWindow setBackgroundColor:[NSColor colorWithDeviceWhite:0.0 alpha:0.0]];
	[zoomWindow setHasShadow:[self hasShadow]];
	[zoomWindow setLevel:NSModalPanelWindowLevel];
	[zoomWindow setOpaque:NO];
	
	imageView = [[NSImageView alloc] initWithFrame:[zoomWindow contentRectForFrameRect:frame]];
	[imageView setImage:image];
	[imageView setImageFrameStyle:NSImageFrameNone];
	[imageView setImageScaling:NSImageScaleAxesIndependently];
	[imageView setAutoresizingMask:NSViewWidthSizable|NSViewHeightSizable];
	
	[zoomWindow setContentView:imageView];
	
	// Reset one shot flag
	[self setOneShot:isOneShot];
	
	__zoomWindow = zoomWindow;
	
	return zoomWindow;
}

- (void)zoomOnFromRect:(NSRect)startRect
{
	NSRect              frame;
	NSWindow            *zoomWindow;

	if ([self isVisible]){
		return;
	}
		
	frame = [self frame];
	
	[self setFrame:frame display:YES];
	
	zoomWindow = [self _createZoomWindowWithRect:startRect];
	[zoomWindow orderFront:self];
	[zoomWindow setFrame:frame display:YES animate:YES];
	
	[self orderFront:nil];
	[zoomWindow close];
	__zoomWindow = nil;
}

-(void)popAway{
	
	NSRect frame = [self frame];
	frame.origin.x += (frame.size.width/2)-10.0;
	frame.origin.y += (frame.size.height/2)-10.0;
	frame.size.width = 20.0;
	frame.size.height = 20.0;
	
	[self zoomOffToRect:frame];
}

-(void)pop{
	[self display];
	
	NSRect frame;
	NSWindow *zoomWindow;
	
	frame = [self frame];
	
	if ([self isVisible]){
		return;
	}
	
	NSRect originalFrame = frame;
	
	NSRect enlargedFrame = originalFrame;
	enlargedFrame.origin.x -= 12.5;
	enlargedFrame.origin.y -= 12.5;
	enlargedFrame.size.width += 25.0;
	enlargedFrame.size.height += 25.0;
	
	NSRect fromRect = originalFrame;
	fromRect.origin.x += originalFrame.size.width/2;
	fromRect.origin.y += originalFrame.size.height/2;
	fromRect.size.width = 1.0;
	fromRect.size.height = 1.0;
	
	zoomWindow = [self _createZoomWindowWithRect:fromRect];
	[zoomWindow orderFront:self];
	[zoomWindow setFrame:enlargedFrame display:YES animate:YES];
	[zoomWindow setFrame:originalFrame display:YES animate:YES];
	
	
	[self makeKeyAndOrderFront:self];	
	[zoomWindow close];
}

- (void)zoomOffToRect:(NSRect)endRect
{
	NSRect              frame;
	NSWindow            *zoomWindow;
	
	frame = [self frame];
	
	if (![self isVisible])
	{
		return;
	}
	
	zoomWindow = [self _createZoomWindowWithRect:frame];
	[zoomWindow orderFront:self];
//	[zoomWindow setUpWindowShadow];
	[self orderOut:self];
	
	
	[zoomWindow setFrame:endRect display:YES animate:YES];
	
	[zoomWindow close];
	
	__zoomWindow = nil;
}

@end

File: XUCore/additions/NSWindow-NoodleEffects.swift
//
//  NSWindow-NoodleEffects.swift
//  XUCore
//
//  Created by Charlie Monroe on 11/21/15.
//  Copyright © 2015 Charlie Monroe Software. All rights reserved.
//

import Foundation

/// Set this variable to a speed-up or slow down the effect
private let kXUZoomAnimationTimeMultiplier = 0.4


private class __XUZoomWindow: NSPanel {
	
	@objc override func animationResizeTime(newWindowFrame: CGRect) -> NSTimeInterval {
		return super.animationResizeTime(newWindowFrame) * kXUZoomAnimationTimeMultiplier
	}
	
}

public extension NSWindow {
	
	private static var __zoomWindow: NSWindow?
	
	/// Creates a new zoom window in screen rect. Nil is returned when there is
	/// no contentView, or the view fails to create the bitmap image representation.
	private func _createZoomWindowWithRect(rect: CGRect) -> NSPanel? {
		let frame = self.frame
		let isOneShot = self.oneShot
		if isOneShot {
			self.oneShot = false
		}
		
		if self.windowNumber <= 0 {
			// Force window device. Kinda crufty but I don't see a visible flash
			// when doing this. May be a timing thing wrt the vertical refresh.
			self.orderBack(self)
			self.orderOut(self)
		}
		
		let image = NSImage(size: frame.size)
		
		guard let view = self.contentView?.superview else {
			return nil
		}
		guard let imageRep = view.bitmapImageRepForCachingDisplayInRect(view.bounds) else {
			return nil
		}
		
		view.cacheDisplayInRect(view.bounds, toBitmapImageRep: imageRep)
		image.addRepresentation(imageRep)
		
		let zoomWindow = __XUZoomWindow(contentRect: rect, styleMask: NSBorderlessWindowMask, backing: .Buffered, `defer`: false)
		zoomWindow.backgroundColor = NSColor(deviceWhite: 0.0, alpha: 0.0)
		zoomWindow.hasShadow = self.hasShadow
		zoomWindow.level = Int(CGWindowLevelForKey(.ModalPanelWindowLevelKey))
		zoomWindow.opaque = false
		
		let imageView = NSImageView(frame: zoomWindow.contentRectForFrameRect(frame))
		imageView.image = image
		imageView.imageFrameStyle = .None
		imageView.imageScaling = .ScaleAxesIndependently
		imageView.autoresizingMask = [ .ViewWidthSizable, .ViewHeightSizable ]
		zoomWindow.contentView = imageView
		
		// Reset one shot flag
		self.oneShot = isOneShot
		
		NSWindow.__zoomWindow = zoomWindow
		
		return zoomWindow
	}
	
	/// Pops the window on screen from startRect.
	public func zoomInFromRect(startRect: CGRect) {
		if self.visible {
			return // Do nothing if we're already on-screen
		}
		
		let frame = self.frame
		self.setFrame(frame, display: true)
		
		let zoomWindow = self._createZoomWindowWithRect(startRect)
		zoomWindow?.orderFront(self)
		zoomWindow?.setFrame(frame, display: true, animate: true)
		
		self.orderFront(nil)
		
		zoomWindow?.close()
		NSWindow.__zoomWindow = nil
	}
	
	/// Removes the window from screen by zooming off to the center of the window.
	public func popAway() {
		var frame = self.frame
		frame.origin.x += (frame.width / 2.0) - 10.0
		frame.origin.y += (frame.size.height / 2.0) - 10.0
		frame.width = 20.0
		frame.size.height = 20.0
		
		self.zoomOutToRect(frame)
	}
	
	/// Pops the window on screen from the middle of self.frame.
	public func pop() {
		self.display()
		
		let frame = self.frame
		if self.visible {
			return // Already visible
		}
		
		let originalFrame = frame
		var enlargedFrame = originalFrame
		enlargedFrame.origin.x -= 12.5
		enlargedFrame.origin.y -= 12.5
		enlargedFrame.width += 25.0
		enlargedFrame.size.height += 25.0
		
		var fromRect = originalFrame
		fromRect.origin.x += originalFrame.width / 2.0
		fromRect.origin.y += originalFrame.size.height / 2.0
		fromRect.width = 1.0
		fromRect.size.height = 1.0
		
		let zoomWindow = self._createZoomWindowWithRect(fromRect)
		zoomWindow?.orderFront(self)
		zoomWindow?.setFrame(enlargedFrame, display: true, animate: true)
		zoomWindow?.setFrame(originalFrame, display: true, animate: true)
		
		self.makeKeyAndOrderFront(self)
		
		zoomWindow?.close()
		
		NSWindow.__zoomWindow = nil
	}
	
	/// Removes the window from screen by zooming off to endRect.
	public func zoomOutToRect(endRect: CGRect) {
		if !self.visible {
			return // Already off screen
		}
		
		let frame = self.frame
		
		let zoomWindow = self._createZoomWindowWithRect(frame)
		zoomWindow?.orderFront(self)
		
		self.orderOut(self)
		
		zoomWindow?.setFrame(endRect, display: true, animate: true)
		zoomWindow?.close()
		
		NSWindow.__zoomWindow = nil
	}
	
}


